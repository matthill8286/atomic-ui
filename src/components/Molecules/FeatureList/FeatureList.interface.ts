export interface Feature {
  name: string
  value: string
  unit?: string | null
}

export interface FeatureListProps {
  features?: Feature[] | null
  showCount?: number
  loading?: boolean
}

export interface FeatureListItemProps {
  name: string
  value: string
  unit?: string | null
}
