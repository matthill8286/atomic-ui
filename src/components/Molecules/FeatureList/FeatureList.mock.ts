export const featureListMockItems = [
  {
    name: 'Ut anim aliquip',
    value: 'Ullamco reprehenderit aliquip',
  },
  {
    name: 'Mollit exercitation',
    value: 'adipisicing magna officia et duis enim commodo non labore ex ex veniam dolor',
  },
  {
    name: 'Reprehenderit et minim',
    value: 'Lorem sunt duis adipisicing',
  },
  {
    name: 'Exercitation est pariatur',
    value: 'Eu dolore fugiat incididunt',
  },
  {
    name: 'Non magna dolor',
    value: 'Minim laborum qui sint qui commodo',
  },
  {
    name: 'Adipisicing culpa',
    value: 'Cupidatat laboris aliqua labore ut',
  },
]
