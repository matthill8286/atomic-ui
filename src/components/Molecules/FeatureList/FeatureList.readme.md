## FeatureList

Product features list, based on names and values
