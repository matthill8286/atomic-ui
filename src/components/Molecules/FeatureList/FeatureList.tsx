import React from 'react'
import { Divider } from '@/components/Atoms/Divider'
import { CopyText } from '@/components/Atoms/Typography'
import { FlexBox } from '@/components/Helper/FlexBox'
import { Cell, Grid, Row } from '@/components/Helper/Grid'
import { styled } from '@/styles'

export type FeatureListType = { label?: string; value: string }

export interface FeatureListProps {
  list: FeatureListType[]
}

export const ListWrapper = styled.div`
  padding-top: ${({ theme }) => theme.spacing.base.sm};
  padding-bottom: ${({ theme }) => theme.spacing.base.sm};
`

export const FeatureList: React.FC<FeatureListProps> = ({ list }): JSX.Element | null => {
  if (!list || !Array.isArray(list) || !list.length) {
    return null
  }

  return (
    <FlexBox display="flex" flexDirection="column">
      {list.map((item, idx) =>
        item.label !== '' && item.label ? (
          <React.Fragment key={`FeatureList-${item.label}-${idx}`}>
            <ListWrapper>
              <Grid noPadding>
                <Row noMargin>
                  <Cell columns={6} colsMd={4} colsSm={4} colsXs={2}>
                    <CopyText weight="semibold" color="grey5" margin="0">
                      {item.label}
                    </CopyText>
                  </Cell>
                  <Cell columns={6} colsMd={4} colsSm={4} colsXs={2}>
                    <CopyText color="grey5" margin="0">
                      {item.value}
                    </CopyText>
                  </Cell>
                </Row>
              </Grid>
            </ListWrapper>
            <Divider color="grey2" height={1} />
          </React.Fragment>
        ) : null
      )}
    </FlexBox>
  )
}

FeatureList.displayName = 'FeatureList'
