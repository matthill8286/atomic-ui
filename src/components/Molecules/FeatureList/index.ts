export * from './FeatureList'
export * from './FeatureList.interface'
