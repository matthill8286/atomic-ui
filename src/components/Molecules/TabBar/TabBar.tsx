import React, { FC, Ref, useEffect, useRef, useState } from 'react'
import { Icon } from '@/components/Atoms/Icon'
import { useWindowDimensions } from '@/components/Helper/useWindowDimensions'
import { StyleguideArrow } from '@excelwithbusiness/webmobile-svg-library'
import { TabBarProps, TabData, TabId } from './TabBar.interface'
import {
  StyledArrowButton,
  StyledTabBar,
  StyledTabsContainer,
  StyledTabBarCard,
} from './TabBar.styled'
import { TabItem } from './TabItem'

export const TabBar: FC<TabBarProps> = props => {
  const {
    onChange,
    selected,
    tabs,
    tabGroupName,
    enableSemanticTheme = false,
    surfaceColor = 'surfaceColor',
    elevation = 0,
    noBorder = false,
    isFixed = false,
    changeDelay = 1000,
    flexDirection,
    ...otherProps
  } = props

  const [tab, setTab] = useState(selected || null)
  const [hasOverflow, setHasOverflow] = useState(false)
  const ref: Ref<HTMLDivElement> = useRef<HTMLDivElement>(null)
  const { width } = useWindowDimensions()

  const handleTabClick = (id: TabId, data?: TabData) => {
    setTab(id)
    if (onChange) {
      onChange(id, data)
    }
  }

  useEffect(() => {
    const handler = setTimeout(() => {
      if (tab !== selected && selected) {
        setTab(selected)
      }
    }, changeDelay)

    return () => {
      clearTimeout(handler)
    }
  }, [selected, changeDelay])

  useEffect(() => {
    if (ref.current) {
      const { offsetWidth, scrollWidth } = ref.current
      const hasWidthOverflow = offsetWidth < scrollWidth

      if (hasOverflow !== hasWidthOverflow) {
        setHasOverflow(hasWidthOverflow)
      }
    }
  }, [tabs, width])

  const handlePrevClick = () => {
    if (ref.current) {
      ref.current.scrollBy({
        behavior: 'smooth',
        left: -200,
      })
    }
  }

  const handleNextClick = () => {
    if (ref.current) {
      ref.current.scrollBy({
        behavior: 'smooth',
        left: 200,
      })
    }
  }

  return (
    <StyledTabBar isFixed={isFixed}>
      <StyledTabBarCard
        elevation={elevation}
        shape="rounded-small"
        noBorder={noBorder ? ['top', 'right', 'bottom', 'left'] : []}
        display="flex"
        surfaceColor={surfaceColor}
        enableSemanticTheme={enableSemanticTheme}
        {...otherProps}>
        {hasOverflow && (
          <StyledArrowButton
            actionType="ghost"
            onClick={handlePrevClick}
            surfaceColor={surfaceColor}
            enableSemanticTheme={enableSemanticTheme}>
            <Icon height={16} width={16} rotate={180}>
              <StyleguideArrow />
            </Icon>
          </StyledArrowButton>
        )}
        <StyledTabsContainer flexDirection={flexDirection} ref={ref}>
          {tabs.map(({ id, label, data, isDisabled = false }) => (
            <TabItem
              key={id}
              id={id}
              label={label}
              data={data}
              tabGroupName={tabGroupName}
              isSelected={tab === id}
              isDisabled={isDisabled}
              onClick={handleTabClick}
              enableSemanticTheme={enableSemanticTheme}
              surfaceColor={surfaceColor}
            />
          ))}
        </StyledTabsContainer>
        {hasOverflow && (
          <StyledArrowButton
            actionType="ghost"
            onClick={handleNextClick}
            enableSemanticTheme={enableSemanticTheme}
            surfaceColor={surfaceColor}>
            <Icon height={16} width={16}>
              <StyleguideArrow />
            </Icon>
          </StyledArrowButton>
        )}
      </StyledTabBarCard>
    </StyledTabBar>
  )
}

TabBar.displayName = 'TabBar'
