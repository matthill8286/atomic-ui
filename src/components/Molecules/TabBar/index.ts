export * from './TabBar'
export * from './TabBar.interface'
export * from './TabItem'
