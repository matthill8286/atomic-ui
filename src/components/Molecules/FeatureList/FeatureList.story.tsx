import { storiesOf } from '@storybook/react'
import { text } from '@storybook/addon-knobs'
import * as React from 'react'
import { FeatureList } from './FeatureList'

storiesOf('Design System/Molecules/FeatureList', module).add('Default', () => (
  <div style={{ width: '400px' }}>
    <FeatureList
      list={[
        { label: text('Provider', 'Provider'), value: 'Youtube' },
        { label: text('Primary competency', 'Primary competency'), value: 'Digital Marketing' },
        { label: text('Type', 'Type'), value: 'Article' },
        { label: text('Length', 'Length'), value: '16 minutes' },
      ]}
    />
  </div>
))
