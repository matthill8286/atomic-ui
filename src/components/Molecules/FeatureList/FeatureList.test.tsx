import React from 'react'
import { mountWithTheme, renderWithTheme } from '@/testRenderer'
import { FeatureList, ListWrapper } from './FeatureList'
import { text } from '@storybook/addon-knobs'
import { TabItem } from '@/components/Molecules/TabBar'

describe('Meta Feature list', () => {
  let inputProps
  beforeEach(() => {
    inputProps = [
      { label: 'Provider', value: 'Youtube' },
      { label: 'Primary competency', value: 'Digital Marketing' },
      { label: 'Type', value: 'Article' },
      { label: 'Length', value: '16 minutes' },
    ]
  })

  it('renders feature list', () => {
    const wrapper = renderWithTheme(<FeatureList list={inputProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders feature list with 3 children', () => {
    const component = mountWithTheme(<FeatureList list={inputProps.slice(0, 3)} />)
    const find = component.find(ListWrapper)
    expect(find.length).toBe(3)
  })

  it('removes the entry with no valid label', () => {
    const listWithMissingLabel = [
      { label: '', value: 'Youtube' },
      { label: 'Type', value: 'Article' },
      { label: 'Primary competency', value: 'Digital Marketing' },
      { label: 'Type', value: 'Article' },
    ]

    const component = mountWithTheme(<FeatureList list={listWithMissingLabel} />)
    const child = component.childAt(0)
    const childProps = child.props().list

    expect(childProps).toStrictEqual(listWithMissingLabel)
    expect(childProps.length).toBe(4)

    const find = component.find(ListWrapper)
    expect(find.length).toBe(3)
  })
})
