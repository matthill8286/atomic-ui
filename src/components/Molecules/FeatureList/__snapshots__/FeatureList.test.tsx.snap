// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meta Feature list renders feature list 1`] = `
.c7 {
  height: 1px;
  width: 100%;
  background: #E5E5E5;
  border-width: 0;
  margin: 0;
}

.c5 {
  margin: 0;
  color: #666666;
  font-size: 14px;
  line-height: 20px;
  font-family: "Oswald","Raleway",Arial,sans-serif;
  font-weight: 600;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c6 {
  margin: 0;
  color: #666666;
  font-size: 14px;
  line-height: 20px;
  font-family: "Oswald","Raleway",Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  box-sizing: border-box;
  margin: calc(16px / 2);
  width: calc(100% - 16px);
  width: calc(50% - 16px);
}

.c2 {
  box-sizing: border-box;
  margin: 0 auto;
  max-width: 1472px;
  padding: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  margin: 0;
  margin-bottom: 0;
}

.c1 {
  padding-top: 16px;
  padding-bottom: 16px;
}

@supports (display:grid) {
  .c4 {
    margin: 0;
    width: auto;
    grid-column-end: span 4;
  }
}

@supports (display:grid) {
  .c4 {
    width: auto;
    grid-column-end: span 2;
  }
}

@media (min-width:512px) {
  .c4 {
    display: block;
    margin: calc(16px / 2);
    width: calc(50% - 16px);
    width: calc(50% - 16px);
  }
}

@media (min-width:752px) {
  .c4 {
    display: block;
    margin: calc(24px / 2);
    width: calc(50% - 24px);
    width: calc(50% - 24px);
  }
}

@media (min-width:1008px) {
  .c4 {
    display: block;
    margin: calc(24px / 2);
    width: calc(33.33333% - 24px);
    width: calc(50% - 24px);
  }
}

@media (min-width:1232px) {
  .c4 {
    display: block;
    margin: calc(24px / 2);
    width: calc(33.33333% - 24px);
    width: calc(50% - 24px);
  }
}

@media screen and (-ms-high-contrast:active),screen and (-ms-high-contrast:none) {

}

@supports (display:grid) {
  .c3 {
    display: grid;
    grid-gap: 16px;
    grid-template-columns: repeat(4,minmax(0,1fr));
  }
}

@media (min-width:512px) {

}

@media (min-width:752px) {

}

@media (min-width:1008px) {

}

@media (min-width:1232px) {

}

<div
  className="c0"
  display="flex"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <p
            className="c5"
            color="grey5"
            fontSize="xs"
          >
            Provider
          </p>
        </div>
        <div
          className="c4"
        >
          <p
            className="c6"
            color="grey5"
            fontSize="xs"
          >
            Youtube
          </p>
        </div>
      </div>
    </div>
  </div>
  <hr
    className="c7"
    color="grey2"
    height={1}
  />
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <p
            className="c5"
            color="grey5"
            fontSize="xs"
          >
            Primary competency
          </p>
        </div>
        <div
          className="c4"
        >
          <p
            className="c6"
            color="grey5"
            fontSize="xs"
          >
            Digital Marketing
          </p>
        </div>
      </div>
    </div>
  </div>
  <hr
    className="c7"
    color="grey2"
    height={1}
  />
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <p
            className="c5"
            color="grey5"
            fontSize="xs"
          >
            Type
          </p>
        </div>
        <div
          className="c4"
        >
          <p
            className="c6"
            color="grey5"
            fontSize="xs"
          >
            Article
          </p>
        </div>
      </div>
    </div>
  </div>
  <hr
    className="c7"
    color="grey2"
    height={1}
  />
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <p
            className="c5"
            color="grey5"
            fontSize="xs"
          >
            Length
          </p>
        </div>
        <div
          className="c4"
        >
          <p
            className="c6"
            color="grey5"
            fontSize="xs"
          >
            16 minutes
          </p>
        </div>
      </div>
    </div>
  </div>
  <hr
    className="c7"
    color="grey2"
    height={1}
  />
</div>
`;
